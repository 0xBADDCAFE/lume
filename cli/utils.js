import { existsSync } from "../deps/fs.js";
import lume from "../mod.js";
import { join, resolve } from "../deps/path.js";
import { brightGreen } from "../deps/colors.js";

/**
 * @return {Promise<*>} a lume instance - ready to build, run, etc.
 */
export async function buildSite(options, site) {
  options.root = resolve(Deno.cwd(), options.root);
  const configFile = join(options.root, options.config);

  if (!site) {
    if (existsSync(configFile)) {
      const mod = await import(`file://${configFile}`);
      site = mod.default;
    } else {
      site = lume();
    }
  }

  site.options.cwd = options.root;

  if (options.dev) {
    site.options.dev = options.dev;
  }

  if (options.location) {
    site.options.location = new URL(options.location);
  }

  if (options.src) {
    site.options.src = options.src;
  }

  if (options.dest) {
    site.options.dest = options.dest;
  }

  if (options["--"]) {
    site.options.flags = options["--"];
  }

  return site;
}

/**
 * @param command name of the command you're validating
 * @param args array of cli arguments (i.e. options._)
 * @param max the max number of args
 * @param min the min number of args
 */
export function validateArgsCount(command, args, max, min = 0) {
  if (args.length > max) {
    throw new Error(`Unexpected arguments: ${args.slice(max).join(", ")}
    
    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }

  if (args.length < min) {
    throw new Error(`Missing arguments

    Run ${brightGreen(`lume ${command} --help`)} for usage information
    `);
  }
}
