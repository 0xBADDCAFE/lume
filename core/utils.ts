import { DOMParser, HTMLDocument } from "../deps/dom.ts";
import {
  brightGreen,
  brightYellow,
  cyan,
  dim,
  green,
  red,
} from "../deps/colors.ts";
import { dirname, extname, join, posix, SEP } from "../deps/path.ts";
import { parse } from "../deps/jsonc.ts";
import { Exception } from "./errors.ts";

/** A list of the available optional plugins */
export const pluginNames = [
  "attributes",
  "base_path",
  "code_highlight",
  "date",
  "esbuild",
  "eta",
  "imagick",
  "inline",
  "jsx",
  "jsx_preact",
  "katex",
  "lightningcss",
  "liquid",
  "metas",
  "minify_html",
  "modify_urls",
  "multilanguage",
  "netlify_cms",
  "on_demand",
  "pagefind",
  "postcss",
  "prism",
  "pug",
  "relations",
  "relative_urls",
  "remark",
  "resolve_urls",
  "sass",
  "sheets",
  "slugify_urls",
  "svgo",
  "terser",
  "windi_css",
];

/** A list of the available plugins with init configurations */
export const initPlugins = [
  "jsx",
  "jsx_preact",
];

export function log(...lines: (string | undefined)[]) {
  console.log("----------------------------------------");
  lines.forEach((line) => line && console.log(line));
  console.log("----------------------------------------");
}

export function promptConfigUpdate(data: unknown) {
  log(
    red("deno.jsonc needs to be manually updated:"),
    dim("Use deno.json to update it automatically"),
    JSON.stringify(data, null, 2),
  );
}

/** Check the compatibility with the current Deno version */
export function checkDenoVersion(): void {
  const minimum = "1.25.4";
  const current = Deno.version.deno;

  if (current < minimum) {
    console.log("----------------------------------------");
    console.error(red("Your Deno version is not compatible with Lume"));
    console.log(`Lume needs Deno ${green(minimum)} or greater`);
    console.log(`Your current version is ${red(current)}`);
    console.log(`Run ${cyan("deno upgrade")} and try again`);
    console.log("----------------------------------------");
    Deno.exit(1);
  }
}

/** Check if the current version is outdated */
export async function checkUpgrade(): Promise<void> {
  const current = getLumeVersion();

  // It's a local version
  if (current.startsWith("local ")) {
    return;
  }

  const stable = !!current.match(/^v\d+\./);
  const expires = 1000 * 60 * 60 * 24; // 1 day
  const interval = localStorage.getItem("lume-upgrade");

  if (interval && parseInt(interval) + expires > Date.now()) {
    return;
  }

  localStorage.setItem("lume-upgrade", Date.now().toString());

  const latest = stable
    ? await getLatestVersion()
    : await getLatestDevelopmentVersion();

  if (current === latest) {
    return;
  }

  let global = false;
  try {
    await Promise.any([
      Deno.stat("deno.json"),
      Deno.stat("deno.jsonc"),
    ]);
    await Deno.stat("import_map.json");
  } catch {
    global = true;
  }

  const command = global
    ? (stable ? "lume upgrade --global" : "lume upgrade --dev --global")
    : (stable ? "deno task lume upgrade" : "deno task lume upgrade --dev");

  log(
    `Update available ${dim(current)}  â†’ ${green(latest)}`,
    `Run ${cyan(command)} to update`,
  );
}

/** Return the latest stable version from the deno.land/x repository */
export async function getLatestVersion(): Promise<string> {
  const response = await fetch("https://cdn.deno.land/lume/meta/versions.json");
  const versions = await response.json();
  return versions.latest;
}

/** Return the hash of the latest commit from the GitHub repository */
export async function getLatestDevelopmentVersion(): Promise<string> {
  const response = await fetch(
    "https://api.github.com/repos/lumeland/lume/commits?per_page=1",
  );
  const commits = await response.json();
  return commits[0].sha;
}

/** Import map file */
export interface ImportMap {
  imports: Record<string, string>;
  scopes?: Record<string, Record<string, string>>;
}

/** Basic options for deno.json file */
export interface DenoConfig {
  importMap?: string;
  tasks?: Record<string, string>;
  compilerOptions?: {
    jsx?: "jsx" | "react-jsx";
    jsxImportSource?: string;
  };
  [key: string]: unknown;
}

/** Run a callback concurrently with all the elements of an Iterable */
export async function concurrent<Type>(
  iterable: AsyncIterable<Type> | Iterable<Type>,
  iteratorFn: (arg: Type) => Promise<unknown>,
  limit = 200,
) {
  const executing: Promise<unknown>[] = [];

  for await (const item of iterable) {
    const p: Promise<unknown> = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

const decoder = new TextDecoder();
const encoder = new TextEncoder();

/** Encode a message using SHA-1 algorithm */
export async function sha1(message: string | Uint8Array): Promise<string> {
  if (typeof message === "string") {
    message = encoder.encode(message);
  }

  const hash = await crypto.subtle.digest("SHA-1", message);
  return decoder.decode(hash);
}

/** Helper to create optional properties recursively */
// deno-lint-ignore ban-types
export type DeepPartial<T> = T extends object ? {
    [P in keyof T]?: DeepPartial<T[P]>;
  }
  : T;

/**
 * Merge two objects recursively.
 * It's used to merge user options with default options.
 */
export function merge<Type>(
  defaults: Type,
  user?: Partial<Type> | DeepPartial<Type>,
): Type {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    if (value === undefined) {
      continue;
    }

    // @ts-ignore: No index signature with a parameter of type 'string' was found on type 'unknown'
    if (isPlainObject(merged[key]) && isPlainObject(value)) {
      // @ts-ignore: Type 'string' cannot be used to index type 'Type'
      merged[key] = merge(merged[key], value);
      continue;
    }

    // @ts-ignore: Type 'string' cannot be used to index type 'Type'
    merged[key] = value;
  }

  return merged;
}

const reactElement = Symbol.for("react.element");
const objectConstructor = {}.constructor;

/** Check if the argument passed is a plain object */
export function isPlainObject(obj: unknown): obj is Record<string, unknown> {
  return typeof obj === "object" && obj !== null &&
    obj.constructor === objectConstructor &&
    // @ts-ignore: Check if the argument passed is a React element
    obj["$$typeof"] !== reactElement;
}

/**
 * Convert the Windows paths (that use the separator "\")
 * to Posix paths (with the separator "/")
 * and ensure it starts with "/".
 */
export function normalizePath(path: string) {
  if (SEP !== "/") {
    path = path.replaceAll(SEP, "/");

    // Is absolute Windows path (C:/...)
    if (path.includes(":/")) {
      return path;
    }
  }

  return posix.join("/", path);
}

/** Convert an HTMLDocument instance to a string */
export function documentToString(document: HTMLDocument) {
  const { doctype, documentElement } = document;

  if (!doctype) {
    return `<!DOCTYPE html>\n${documentElement?.outerHTML || ""}`;
  }

  return `<!DOCTYPE ${doctype.name}` +
    (doctype.publicId ? ` PUBLIC "${doctype.publicId}"` : "") +
    (!doctype.publicId && doctype.systemId ? " SYSTEM" : "") +
    (doctype.systemId ? ` "${doctype.systemId}"` : "") +
    `>\n${documentElement?.outerHTML}`;
}

const parser = new DOMParser();

/** Parse a string with HTML code and return an HTMLDocument */
export function stringToDocument(string: string): HTMLDocument {
  const document = parser.parseFromString(string, "text/html");

  if (!document) {
    throw new Error("Unable to parse the HTML code");
  }

  return document;
}

/** Return the current installed version */
export function getLumeVersion(
  url = new URL(import.meta.resolve("../")),
): string {
  const { pathname } = url;
  return pathname.match(/@([^/]+)/)?.[1] ?? `local (${pathname})`;
}

/** Returns the _config file of a site */
export async function getConfigFile(
  path?: string,
): Promise<string | undefined> {
  if (path) {
    try {
      return await Deno.realPath(path);
    } catch (cause) {
      throw new Exception("Config file not found", { path, cause });
    }
  }

  const paths = ["_config.js", "_config.ts"];

  for (const path of paths) {
    try {
      return await Deno.realPath(path);
    } catch {
      // Ignore
    }
  }
}

export interface DenoConfigResult {
  file: string;
  config: DenoConfig;
  importMap?: ImportMap;
}

/** Detect and returns the Deno configuration */
export async function readDenoConfig(
  importMapFile?: string,
): Promise<DenoConfigResult | undefined> {
  for (const file of ["deno.json", "deno.jsonc"]) {
    try {
      const content = await Deno.readTextFile(file);
      const config = parse(content) as DenoConfig;
      importMapFile ||= config.importMap;
      if (importMapFile) {
        config.importMap = importMapFile;
        const importMap: ImportMap = isUrl(importMapFile)
          ? await (await fetch(importMapFile)).json()
          : await JSON.parse(await Deno.readTextFile(importMapFile));

        return { file, config, importMap };
      }
      return { file, config };
    } catch (err) {
      if (err instanceof Deno.errors.NotFound) {
        continue;
      }

      throw err;
    }
  }
}

/** Update the Deno configuration */
export async function writeDenoConfig(options: DenoConfigResult) {
  const { file, config, importMap } = options;

  if (importMap && !config.importMap) {
    config.importMap = "./import_map.json";
  }

  if (config.importMap) {
    const importMapFile = join(dirname(file), config.importMap);
    console.log(importMapFile);

    await Deno.writeTextFile(
      importMapFile,
      JSON.stringify(importMap, null, 2) + "\n",
    );
    console.log(brightGreen("Import map file saved:"), importMapFile);
  }

  if (extname(file) === ".jsonc") {
    console.log(
      brightYellow("deno.jsonc needs to be manually updated:"),
      "Use deno.json to update it automatically",
      JSON.stringify(config, null, 2),
    );
  } else {
    await Deno.writeTextFile(file, JSON.stringify(config, null, 2) + "\n");
    console.log(brightGreen("Deno configuration file saved:"), file);
  }
}

export function isUrl(path: string): boolean {
  return !!path.match(/^(https?|file):\/\//);
}

export function createDate(str: string | number): Date | undefined {
  if (typeof str === "number") {
    return new Date(str);
  }

  const datetime = str.match(
    /^(\d{4})-(\d\d)-(\d\d)(?:-(\d\d)-(\d\d)(?:-(\d\d))?)?$/,
  );

  if (datetime) {
    const [, year, month, day, hour, minute, second] = datetime;

    return new Date(Date.UTC(
      parseInt(year),
      parseInt(month) - 1,
      parseInt(day),
      hour ? parseInt(hour) : 0,
      minute ? parseInt(minute) : 0,
      second ? parseInt(second) : 0,
    ));
  }

  if (str.match(/^\d+$/)) {
    return new Date(parseInt(str));
  }
}

export async function read(path: string, isBinary: true): Promise<Uint8Array>;
export async function read(path: string, isBinary: false): Promise<string>;
export async function read(
  path: string,
  isBinary: boolean,
): Promise<string | Uint8Array> {
  if (!isUrl(path)) {
    return isBinary ? Deno.readFile(path) : Deno.readTextFile(path);
  }

  const response = await fetch(path);

  return isBinary
    ? new Uint8Array(await response.arrayBuffer())
    : response.text();
}
