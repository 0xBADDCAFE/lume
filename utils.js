import { bold, red } from "./deps/colors.js";
import { SEP } from "./deps/path.js";

export async function concurrent(iterable, iteratorFn, limit = 200) {
  const executing = [];

  for await (const item of iterable) {
    const p = iteratorFn(item).then(() =>
      executing.splice(executing.indexOf(p), 1)
    );

    executing.push(p);

    if (executing.length >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
}

export const mimes = new Map([
  [".html", "text/html; charset=utf-8"],
  [".txt", "text/plain; charset=utf-8"],
  [".js", "text/javascript; charset=utf-8"],
  [".css", "text/css; charset=utf-8"],
  [".atom", "application/atom+xml; charset=utf-8"],
  [".rss", "application/rss+xml; charset=utf-8"],
  [".svg", "image/svg+xml; charset=utf-8"],
  [".json", "application/json"],
  [".webmanifest", "application/manifest+json"],
  [".ico", "image/x-icon"],
  [".avif", "image/avif"],
  [".png", "image/png"],
  [".apng", "image/apng"],
  [".jpg", "image/jpg"],
  [".jpeg", "image/jpg"],
  [".gif", "image/gif"],
  [".mp3", "audio/mpeg"],
  [".mp4", "video/mpeg"],
  [".xml", "text/xml"],
  [".woff", "font/woff"],
  [".woff2", "font/woff2"],
  [".wasm", "application/wasm"],
  [".webp", "image/webp"],
  [".webm", "video/webm"],
  [".zip", "application/zip"],
]);

export function error(context, message, exception) {
  console.error(bold(red(`${context}:`)), message);
  if (exception) {
    console.error(exception);
  }
  console.error("");
}

export function merge(defaults, user) {
  const merged = { ...defaults };

  if (!user) {
    return merged;
  }

  for (const [key, value] of Object.entries(user)) {
    if (
      typeof merged[key] === "object" && typeof value === "object" &&
      !Array.isArray(merged[key])
    ) {
      merged[key] = merge(merged[key], value);
      continue;
    }

    merged[key] = value;
  }

  return merged;
}

export function normalizePath(path) {
  // Is Windows path
  if (SEP !== "/") {
    return path.replaceAll(SEP, "/");
  }

  return path;
}

export const slugChars = new Map([
  ["รฐ", "d"],
  ["ร", "ss"],
]);

export function slugify(string) {
  return string.toLowerCase()
    .normalize("NFKD")
    .replaceAll(/[\s_-]+/g, "-")
    .replaceAll(/[^\w\/-]/g, (char) => slugChars.get(char) || "");
}

export function searchByExtension(path, extensions) {
  for (const [key, value] of extensions) {
    if (path.endsWith(key)) {
      return [key, value];
    }
  }
}
