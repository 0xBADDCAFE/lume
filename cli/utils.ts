import lume from "../mod.ts";
import { toFileUrl } from "../deps/path.ts";
import { dim, red } from "../deps/colors.ts";
import { getConfigFile } from "../core/utils.ts";

import type { Site } from "../core.ts";

/** Create a site instance */
export async function createSite(root: string, config?: string): Promise<Site> {
  const path = await getConfigFile(root, config);

  if (path) {
    console.log(`Loading config file ${dim(path)}`);
    console.log();
    const mod = await import(toFileUrl(path).href);
    return mod.default;
  }

  return lume();
}

/** A list of the available optional plugins */
export const pluginNames = [
  "attributes",
  "base_path",
  "code_highlight",
  "date",
  "esbuild",
  "eta",
  "imagick",
  "inline",
  "jsx",
  "lightningcss",
  "liquid",
  "metas",
  "minify_html",
  "modify_urls",
  "multilanguage",
  "nano_jsx",
  "netlify_cms",
  "on_demand",
  "postcss",
  "prism",
  "pug",
  "relations",
  "relative_urls",
  "resolve_urls",
  "sass",
  "slugify_urls",
  "svgo",
  "terser",
  "windi_css",
];

export function log(...lines: (string | undefined)[]) {
  console.log("----------------------------------------");
  lines.forEach((line) => line && console.log(line));
  console.log("----------------------------------------");
}

export function promptConfigUpdate(data: unknown) {
  log(
    red("deno.jsonc needs to be manually updated:"),
    dim("Use deno.json to update it automatically"),
    JSON.stringify(data, null, 2),
  );
}
