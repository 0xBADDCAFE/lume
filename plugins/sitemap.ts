import { merge } from "../core/utils.ts";
import { Page } from "../core/filesystem.ts";
import { Search } from "../plugins/search.ts";

import type { Data, Site, StaticFile } from "../core.ts";

export interface Options {
  /** The sitemap file name */
  filename: string;

  /** The query to search pages included in the sitemap */
  query: string;

  /** The values to sort the sitemap */
  sort: string;
}

// Default options
export const defaults: Options = {
  filename: "/sitemap.xml",
  query: "",
  sort: "url=asc",
};

/** A plugin to generate a sitemap.xml from page files after build */
export default function (userOptions?: Partial<Options>) {
  const options = merge(defaults, userOptions);

  return (site: Site) => {
    site.addEventListener("afterRender", () => {
      // Create the sitemap.xml page
      const sitemap = Page.create(options.filename, getSitemapContent(site));

      // Add to the sitemap page to pages
      site.pages.push(sitemap);

      // Search for the `robots.txt` file
      const robots = site.files.some((file: StaticFile) =>
        file.dest === "/robots.txt"
      );

      // If the `robots.txt` file doesn't exist, create it
      if (!robots) {
        const robots = site.pages.find((page: Page) =>
          page.data.url === "/robots.txt"
        );

        if (robots) {
          robots.content += `Sitemap: ${site.url(options.filename, true)}`;
        } else {
          site.pages.push(Page.create(
            "/robots.txt",
            `User-agent: *\nAllow: /\n\nSitemap: ${
              site.url("/sitemap.xml", true)
            }`,
          ));
        }
      }
    });

    function getSitemapContent(site: Site) {
      const search = new Search(site, true);
      const sitemap = search.pages(options.query, options.sort) as Data[];

      // deno-fmt-ignore
      return `
<?xml version="1.0" encoding="utf-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
  ${sitemap.map((data: Data) =>
  `<url>
    <loc>${site.url(data.url as string, true)}</loc>
    ${data.date ? `<lastmod>${data.date.toISOString()}</lastmod>` : ""}
  </url>
  `).join("").trim()}
</urlset>`.trim();
    }
  };
}
