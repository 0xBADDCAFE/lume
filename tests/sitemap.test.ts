import { assertSiteSnapshot, build, getSite } from "./utils.ts";
import sitemap from "../plugins/sitemap.ts";

Deno.test("Sitemap plugin", async (t) => {
  const site = getSite({
    src: "normal",
    server: {
      page404: "/page5/",
    },
  });

  site.use(sitemap());
  site.ignore("static.yml");

  await build(site);
  await assertSiteSnapshot(t, site);
});
