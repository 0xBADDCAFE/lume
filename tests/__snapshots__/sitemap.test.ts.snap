export const snapshot = {};

snapshot[`Sitemap plugin 1`] = `8`;

snapshot[`Sitemap plugin 2`] = `
{
  formats: [
    {
      engines: 1,
      ext: ".tmpl.ts",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: 1,
      ext: ".tmpl.js",
      pageLoader: [AsyncFunction: module],
    },
    {
      engines: undefined,
      ext: ".tmpl.json",
      pageLoader: [AsyncFunction: json],
    },
    {
      dataLoader: [AsyncFunction: json],
      engines: undefined,
      ext: ".json",
    },
    {
      engines: 1,
      ext: ".md",
      pageLoader: [AsyncFunction: text],
    },
    {
      engines: 1,
      ext: ".markdown",
      pageLoader: [AsyncFunction: text],
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".js",
    },
    {
      componentLoader: [AsyncFunction: module],
      dataLoader: [AsyncFunction: module],
      engines: 1,
      ext: ".ts",
    },
    {
      componentLoader: [AsyncFunction: text],
      engines: 1,
      ext: ".njk",
      includesPath: "_includes",
      pageLoader: [AsyncFunction: text],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yaml",
      pageLoader: [AsyncFunction: yaml],
    },
    {
      dataLoader: [AsyncFunction: yaml],
      engines: undefined,
      ext: ".yml",
      pageLoader: [AsyncFunction: yaml],
    },
  ],
}
`;

snapshot[`Sitemap plugin 3`] = `
[
]
`;

snapshot[`Sitemap plugin 4`] = `
{
  content: "User-agent: *
Allow: /

Sitemap: https://example.com/sitemap.xml",
  data: {
    children: false,
    content: "User-agent: *
Allow: /

Sitemap: https://example.com/sitemap.xml",
    page: undefined,
    url: "/robots.txt",
  },
  dest: {
    ext: ".txt",
    path: "/robots",
  },
  src: {
    asset: true,
    path: "",
    remote: undefined,
    slug: "robots",
  },
}
`;

snapshot[`Sitemap plugin 5`] = `
{
  content: '<?xml version="1.0" encoding="utf-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
  <url>
    <loc>https://example.com/overrided-page2/</loc>
    <lastmod>2020-06-21T00:00:00.000Z</lastmod>
  </url>
  <url>
    <loc>https://example.com/pages/page4/</loc>
    <lastmod>2021-01-02T18:32:00.000Z</lastmod>
  </url>
  <url>
    <loc>https://example.com/pages/page6/</loc>
    <lastmod>2022-01-01T00:00:00.000Z</lastmod>
  </url>
  <url>
    <loc>https://example.com/pages/subpage/page7/</loc>
    <lastmod>2022-01-01T00:00:00.000Z</lastmod>
  </url>
</urlset>',
  data: {
    children: false,
    content: '<?xml version="1.0" encoding="utf-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
  <url>
    <loc>https://example.com/overrided-page2/</loc>
    <lastmod>2020-06-21T00:00:00.000Z</lastmod>
  </url>
  <url>
    <loc>https://example.com/pages/page4/</loc>
    <lastmod>2021-01-02T18:32:00.000Z</lastmod>
  </url>
  <url>
    <loc>https://example.com/pages/page6/</loc>
    <lastmod>2022-01-01T00:00:00.000Z</lastmod>
  </url>
  <url>
    <loc>https://example.com/pages/subpage/page7/</loc>
    <lastmod>2022-01-01T00:00:00.000Z</lastmod>
  </url>
</urlset>',
    page: undefined,
    url: "/sitemap.xml",
  },
  dest: {
    ext: ".xml",
    path: "/sitemap",
  },
  src: {
    asset: true,
    path: "",
    remote: undefined,
    slug: "sitemap",
  },
}
`;

snapshot[`Sitemap plugin 6`] = `
{
  content: "<!DOCTYPE html>
Content of Page 5",
  data: {
    children: true,
    content: "Content of Page 5",
    date: 1970-01-01T00:00:00.000Z,
    imagick: [
      {
        resize: [
          400,
          300,
        ],
        suffix: "-small",
      },
    ],
    mergedKeys: {
      imagick: "array",
      metas: "object",
      tags: "stringArray",
    },
    metas: {
      description: "Default description",
      title: "Default title",
    },
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    site: "Default site name",
    tags: [
    ],
    title: "Page 5",
    url: "/page5/",
  },
  dest: {
    ext: ".html",
    path: "/page5/index",
  },
  src: {
    asset: undefined,
    ext: ".yaml",
    path: "/page5",
    remote: undefined,
    slug: "page5",
  },
}
`;

snapshot[`Sitemap plugin 7`] = `
{
  content: "<!DOCTYPE html>
Content of Page 2",
  data: {
    children: true,
    colors: [
      "#ff0000",
      "#0000ff",
      "#00ff00",
    ],
    content: "Content of Page 2",
    date: 1970-01-01T00:00:00.000Z,
    documents: [
      {
        title: "Document 1",
      },
      {
        title: "Document 2",
      },
      {
        title: "Document 3",
      },
    ],
    drinks: {
      alcoholic: [
        "gin-tonic",
        "gin-soda",
      ],
      others: [
        "coffee",
        "tea",
      ],
    },
    imagick: [
      {
        resize: [
          400,
          300,
        ],
        suffix: "-small",
      },
    ],
    mergedKeys: {
      imagick: "array",
      metas: "object",
      tags: "stringArray",
    },
    metas: {
      description: "Default description",
      title: "Default title",
    },
    names: [
      "Oscar",
      "Laura",
    ],
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    site: "Folder overrided site name",
    tags: [
      "sub-pages",
    ],
    title: "Page 2",
    url: "/overrided-page2/",
  },
  dest: {
    ext: ".html",
    path: "/overrided-page2/index",
  },
  src: {
    asset: undefined,
    ext: ".tmpl.json",
    path: "/pages/2020-06-21_page2",
    remote: undefined,
    slug: "page2",
  },
}
`;

snapshot[`Sitemap plugin 8`] = `
{
  content: "<!DOCTYPE html>
Content of Page 4 in Overrided site name, from the file /pages/2021-01-02-18-32_page4.tmpl.ts",
  data: {
    children: true,
    colors: [
      "#ff0000",
      "#0000ff",
      "#00ff00",
    ],
    content: undefined,
    date: 1970-01-01T00:00:00.000Z,
    documents: [
      {
        title: "Document 1",
      },
      {
        title: "Document 2",
      },
      {
        title: "Document 3",
      },
    ],
    drinks: {
      alcoholic: [
        "gin-tonic",
        "gin-soda",
      ],
      others: [
        "coffee",
        "tea",
      ],
    },
    imagick: [
      {
        resize: [
          400,
          300,
        ],
        suffix: "-small",
      },
    ],
    mergedKeys: {
      imagick: "array",
      metas: "object",
      tags: "stringArray",
    },
    metas: {
      description: "Default description",
      title: "Default title",
    },
    names: [
      "Oscar",
      "Laura",
    ],
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    site: "Overrided site name",
    tags: [
      "sub-pages",
    ],
    title: "Page 4",
    url: "/pages/page4/",
  },
  dest: {
    ext: ".html",
    path: "/pages/page4/index",
  },
  src: {
    asset: undefined,
    ext: ".tmpl.ts",
    path: "/pages/2021-01-02-18-32_page4",
    remote: undefined,
    slug: "page4",
  },
}
`;

snapshot[`Sitemap plugin 9`] = `
{
  content: "<!DOCTYPE html>
<p>Content of Page 6</p>
",
  data: {
    children: true,
    colors: [
      "#ff0000",
      "#0000ff",
      "#00ff00",
    ],
    content: "Content of Page 6
",
    date: 1970-01-01T00:00:00.000Z,
    documents: [
      {
        title: "Document 1",
      },
      {
        title: "Document 2",
      },
      {
        title: "Document 3",
      },
    ],
    drinks: {
      alcoholic: [
        "gin-tonic",
        "gin-soda",
      ],
      others: [
        "coffee",
        "tea",
      ],
    },
    imagick: [
      {
        resize: [
          400,
          300,
        ],
        suffix: "-small",
      },
    ],
    mergedKeys: {
      imagick: "array",
      metas: "object",
      tags: "stringArray",
    },
    metas: {
      description: "Default description",
      title: "Default title",
    },
    names: [
      "Oscar",
      "Laura",
    ],
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    site: "Folder overrided site name",
    tags: [
      "sub-pages",
      "page6",
    ],
    title: "Page 6",
    url: "/pages/page6/",
  },
  dest: {
    ext: ".html",
    path: "/pages/page6/index",
  },
  src: {
    asset: undefined,
    ext: ".md",
    path: "/pages/2021-12-29-page6",
    remote: undefined,
    slug: "page6",
  },
}
`;

snapshot[`Sitemap plugin 10`] = `
{
  content: "Content of Page 3",
  data: {
    children: true,
    colors: [
      "#ff0000",
      "#0000ff",
      "#00ff00",
    ],
    content: "Content of Page 3",
    date: 1970-01-01T00:00:00.000Z,
    documents: [
      {
        title: "Document 1",
      },
      {
        title: "Document 2",
      },
      {
        title: "Document 3",
      },
    ],
    drinks: {
      alcoholic: [
        "gin-tonic",
        "gin-soda",
      ],
      others: [
        "coffee",
        "tea",
      ],
    },
    imagick: [
      {
        resize: [
          400,
          300,
        ],
        suffix: "-small",
      },
    ],
    mergedKeys: {
      imagick: "array",
      metas: "object",
      tags: "stringArray",
    },
    metas: {
      description: "Default description",
      title: "Default title",
    },
    names: [
      "Oscar",
      "Laura",
    ],
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    site: "Folder overrided site name",
    tags: [
      "sub-pages",
    ],
    title: "Page 3",
    url: "/page_3",
  },
  dest: {
    ext: "",
    path: "/page_3",
  },
  src: {
    asset: undefined,
    ext: ".tmpl.js",
    path: "/pages/page3",
    remote: undefined,
    slug: "page3",
  },
}
`;

snapshot[`Sitemap plugin 11`] = `
{
  content: "<!DOCTYPE html>
Content of Page 7",
  data: {
    children: true,
    colors: [
      "#ff0000",
      "#0000ff",
      "#00ff00",
    ],
    content: "Content of Page 7",
    date: 1970-01-01T00:00:00.000Z,
    documents: [
      {
        title: "Document 1",
      },
      {
        title: "Document 2",
      },
      {
        title: "Document 3",
      },
    ],
    drinks: {
      alcoholic: [
        "gin-tonic",
        "gin-soda",
      ],
      others: [
        "coffee",
        "tea",
      ],
    },
    imagick: [
      {
        resize: [
          400,
          300,
        ],
        suffix: "-small",
      },
    ],
    mergedKeys: {
      imagick: "array",
      metas: "object",
      tags: "stringArray",
    },
    metas: {
      description: "Default description",
      title: "Default title",
    },
    names: [
      "Oscar",
      "Laura",
    ],
    page: undefined,
    paginate: [Function: paginate],
    search: Search {},
    site: "Folder overrided site name",
    tags: [
      "sub-pages",
      "sub-sub-pages",
      "page7",
    ],
    url: "/pages/subpage/page7/",
  },
  dest: {
    ext: ".html",
    path: "/pages/subpage/page7/index",
  },
  src: {
    asset: undefined,
    ext: ".tmpl.js",
    path: "/pages/subpage/page7",
    remote: undefined,
    slug: "page7",
  },
}
`;
